spring-web/src/main/java/org/springframework/http/client/JdkClientHttpRequest.java

```diff
@@ -48,8 +51,19 @@ class JdkClientHttpRequest extends AbstractStreamingClientHttpRequest {
 	 * The JDK HttpRequest doesn't allow all headers to be set. The named headers are taken from the default
 	 * implementation for HttpRequest.
 	 */
-	private static final List<String> DISALLOWED_HEADERS =
-			List.of("connection", "content-length", "expect", "host", "upgrade");
+	protected static final Set<String> DISALLOWED_HEADERS = getDisallowedHeaders();
+
+	private static Set<String> getDisallowedHeaders() {
+		TreeSet<String> headers = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+		headers.addAll(Set.of("connection", "content-length", "expect", "host", "upgrade"));
+
+		String headersToAllow = System.getProperty("jdk.httpclient.allowRestrictedHeaders");
```

Comment:
Maybe there is a setter somewhere and we set the property rather than trying to read it?